beets/metadata_plugins.py

```diff
@@ -0,0 +1,352 @@
+"""Metadata source plugin interface.
+
+This allows beets to lookup metadata from various sources. We define
+a common interface for all metadata sources which need to be
+implemented as plugins.
+"""
+
+from __future__ import annotations
+
+import abc
+import re
+from typing import TYPE_CHECKING, Generic, Literal, Sequence, TypedDict, TypeVar
+
+from typing_extensions import NotRequired
+
+from .plugins import BeetsPlugin, find_plugins, notify_info_yielded, send
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+    from confuse import ConfigView
+
+    from .autotag import Distance
+    from .autotag.hooks import AlbumInfo, Item, TrackInfo
+
+
+def find_metadata_source_plugins() -> list[MetadataSourcePlugin]:
+    """Returns a list of MetadataSourcePluginNew subclass instances from all
+    currently loaded beets plugins.
+    """
+    return [
+        plugin
+        for plugin in find_plugins()
+        if isinstance(plugin, MetadataSourcePlugin)
+    ]
+
+
+@notify_info_yielded("albuminfo_received")
+def candidates(*args, **kwargs) -> Iterable[AlbumInfo]:
+    """Return matching album candidates by using all metadata source
+    plugins."""
+    for plugin in find_metadata_source_plugins():
+        yield from plugin.candidates(*args, **kwargs)
+
+
+@notify_info_yielded("trackinfo_received")
+def item_candidates(*args, **kwargs) -> Iterable[TrackInfo]:
+    """Return matching track candidates by using all metadata source
+    plugins."""
+    for plugin in find_metadata_source_plugins():
+        yield from plugin.item_candidates(*args, **kwargs)
+
+
+def album_for_id(_id: str) -> AlbumInfo | None:
+    """Get AlbumInfo object for the given ID string.
+
+    A single ID can yield just a single album, so we return the first match.
+    """
+    for plugin in find_metadata_source_plugins():
+        if info := plugin.album_for_id(album_id=_id):
+            send("albuminfo_received", info=info)
+            return info
+
+    return None
+
+
+def track_for_id(_id: str) -> TrackInfo | None:
+    """Get TrackInfo object for the given ID string.
+
+    A single ID can yield just a single track, so we return the first match.
+    """
+    for plugin in find_metadata_source_plugins():
+        if info := plugin.track_for_id(_id):
+            send("trackinfo_received", info=info)
+            return info
+
+    return None
```

Comment:
```suggestion
        yield from plugins.tracks_for_ids([_id])
```

I'd say we can move the logging message under `tracks_for_ids`:
1. Reuse `None` filtering logic in there
2. Gives an opportunity to include the plugin name in the log message (which I've personally been missing).
3. Should be able to drop `notify_info_yielded` from around `def candidates/item_candidates` above

Reply:
We could do this, I see the following issue: in theory the plugin implementation can overwrite these functions and ensuring that the send is always called will be more difficult (calling by_id vs by_ids). Also it might be introduce duplication of code.

We can always include the plugin name in the send, something like this could also work:

```python
        if info := plugin.album_for_id(album_id=_id):
            send("albuminfo_received", info=info, data_source=plugin.data_source)
            return info
```

Reply:
Makes sense